### PR TITLE
Fix 404 when segment is used but not in the Coordinator snapshot

### PR DESCRIPTION
**Pull request by AmatyaAvadhanula**
Fixes the issue where a used segment is available in the metadata store but not in the Coordinator's snapshot

### Description

An index_parallel task may fetch used segment ids from the metadata store, but its subtasks may fetch the relevant payloads again from the Coordinator's snapshot during ingestion. This snapshot can be stale and lead to a 404 when the segment is absent from the snapshot.

MetadataResource#getUsedSegment now relies on the Coordinator snapshot and then falls back to the metadata store to fix the occurrence of segment not found.

<hr>

##### Key changed/added classes in this PR
 * `MetadataResource`
 *  `IndexerStorageMetadataCoordinator`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by gianm** [comment]
@AmatyaAvadhanula please spell out which kind of task you have in mind for this issue. It helps people understand what the bug is.

I am aware of an similar (but different) issue with MSQ tasks, where the controller uses `/druid/coordinator/v1/metadata/datasources/{dataSource}/segments?full` to get a list of used segments, and the workers use `/druid/coordinator/v1/metadata/datasources/{dataSource}/segments/{segmentId}`. However both of these are implemented in `MetadataResource` using `segmentsMetadataManager.getImmutableDataSourceWithUsedSegments(dataSourceName)`, so they would both use the snapshot, and the issue you describe shouldn't happen. The similar, but different, issue is that there could be a compaction or `REPLACE` operation running between these two calls, so a segment that was used in the first call could become unused by the second call. This patch would not fix that issue.

**Comment by cryptoe** [comment]
@gianm MSQ uses the post version of the API : https://github.com/apache/druid/blob/3ddcaa0180ec58ae1476f3d54610c30ace3a948e/server/src/main/java/org/apache/druid/client/coordinator/CoordinatorClientImpl.java#L74-L75

which uses the metadataStorageConnector: 
https://github.com/apache/druid/blob/3ddcaa0180ec58ae1476f3d54610c30ace3a948e/server/src/main/java/org/apache/druid/server/http/MetadataResource.java#L267-L268

So I guess this fix is okay unless I am missing something.

File: server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java:218]: Nit: Should we mark this as nullable.

**Review by cryptoe — APPROVED** [review]
Changes LGTM. Please wait for @gianm approval before we go ahead and merge this.

**Pull request merged**
